Subject: generate.py: get it working with PyYAML 5.3.1 and Jinja2 2.11.2
Forwarded: https://github.com/justinian/j6-uefi-headers/pull/1
From: TK Chia <tkchia@users.noreply.github.com>
Last-Update: 2020-12-18

diff --git a/generate.py b/generate.py
index 259e0c7..3825c8c 100755
--- a/generate.py
+++ b/generate.py
@@ -17,7 +17,7 @@ os.makedirs(protos_dir, exist_ok=True)
 proto_template = env.get_template("proto.j2")
 
 import yaml
-protos = yaml.load(open("protos.yaml"))
+protos = yaml.safe_load(open("protos.yaml"))
 for proto in protos:
     with open(os.path.join(protos_dir, proto["name"] + ".h"), 'w') as header:
-        header.write(template.render(proto_template))
+        header.write(proto_template.render(proto))
